### PR TITLE
[Form] Add placeholder as option in all TextType(TextareaType, EmailType,...)

### LINKED ISSUE #45214: Support option 'placeholder' for all relevant form field types like 'TextType'
### Description

It's confusing, that "ChoiceType" does support the option "placeholder", but other types like "TextType" do not.

Currently placeholder has to be given via 'attr' option.

### Example

**Currently**

```
$builder->add('location', TextType::class, [
    'attr' => ['placeholder' => 'Town or State)'],
]);
```

**Proposed:**

```
$builder->add('location', TextType::class, [
    'placeholder' => 'Town or State',
]);
```

### PR DESCRIPTION
**Pull request by johdcode**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | close #45214
| License       | MIT
| Doc PR        | -

### CODE REVIEW

**Comment by johdcode** [comment]
Tests added !

**Comment by johdcode** [comment]
Should i create a new PR for a better history ?

**Comment by alexislefebvre** [comment]
> Should i create a new PR for a better history ?

There's no need to alter the commits, they will be squashed when the PR will be merged.

File: src/Symfony/Component/Form/Extension/Core/Type/TextType.php [hunk 1]
```diff
@@ -30,13 +32,24 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+    }
+
     /**
      * {@inheritdoc}
      */
     public function configureOptions(OptionsResolver $resolver)
     {
         $resolver->setDefaults([
             'compound' => false,
+            'placeholder' => '',
```
Comment (GromNaN) [inline src/Symfony/Component/Form/Extension/Core/Type/TextType.php:52]: You can restrict type with `$resolver->setAllowedTypes('placeholder', 'string');`

File: src/Symfony/Component/Form/Extension/Core/Type/TextType.php [hunk 2]
```diff
@@ -30,13 +32,24 @@ public function buildForm(FormBuilderInterface $builder, array $options)
         }
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    public function buildView(FormView $view, FormInterface $form, array $options)
+    {
+        if (isset($options['placeholder']) && '' != $options['placeholder']) {
```
Comment (GromNaN) [inline src/Symfony/Component/Form/Extension/Core/Type/TextType.php]: Strict comparison is more safe.

```suggestion
        if (isset($options['placeholder']) && '' !== $options['placeholder']) {
```

File: src/Symfony/Component/Form/Test/TypeTestCase.php [hunk 3]
```diff
@@ -27,6 +28,8 @@ abstract class TypeTestCase extends FormIntegrationTestCase
      */
     protected $dispatcher;
 
+    public const PLACEHOLDER_OPTION_TEXT = 'My placeholder...';
```
Comment (GromNaN) [inline src/Symfony/Component/Form/Test/TypeTestCase.php]: It's used only inside this class. Should be `private`.

File: src/Symfony/Component/Form/Test/TypeTestCase.php [hunk 4]
```diff
@@ -62,4 +65,112 @@ public static function assertDateIntervalEquals(\DateInterval $expected, \DateIn
     {
         self::assertEquals($expected->format('%RP%yY%mM%dDT%hH%iM%sS'), $actual->format('%RP%yY%mM%dDT%hH%iM%sS'));
     }
+
+    public function testTextTypePlaceholderOption()
```
Comment (GromNaN) [inline src/Symfony/Component/Form/Test/TypeTestCase.php]: The class `TypeTestCase` is extended by several test classes; consequently test methods are run for each child class. That's waste of machine time.

The assertion for each form type should be in the related test class (`TypeTestCase::testTextTypePlaceholderOption` become `TextTypeTest::testPlaceholderOption`)

**Review by GromNaN — CHANGES_REQUESTED** [review]
Thanks for working on this useful feature. I have some observations: the tests should be reworked.

**Comment by johdcode** [comment]
With pleasure, I tried to find one useful but accessible for my first PR, and thanks for your really clear explains I work it

**Comment by johdcode** [comment]
@GromNaN should you have any other suggestions for me ?

**Comment by stof** [comment]
Closing this PR for the reason given by @xabbuh on the issue: https://github.com/symfony/symfony/issues/45214#issuecomment-1096650865

**Pull request closed without merging**